drivers/display/display_ili9341.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2020 Teslabs Engineering S.L.
+ * Copyright (c) 2021 Krivorot Oleg <krivorot.oleg@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "display_ili9341.h"
+#include "display_ili9xxx.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(display_ili9341, CONFIG_DISPLAY_LOG_LEVEL);
+
+int ili9341_regs_init(const struct device *dev)
+{
+	const struct ili9xxx_config *config = dev->config;
+	const struct ili9341_regs *regs = config->regs;
+
+	int r;
+	uint8_t tmp[1] = { ILI9341_SWRESET };
+	LOG_HEXDUMP_DBG(tmp, 1, "SW RESET");
+	r = ili9xxx_transmit(dev, ILI9341_SWRESET, NULL, 0);
+	if (r < 0) {
+		return r;
+	}
+	k_sleep(K_MSEC(5));
+
+	tmp[0] = ILI9XXX_DISPOFF;
+	LOG_HEXDUMP_DBG(tmp, 1, "DISPOFF");
+	r = ili9xxx_transmit(dev, ILI9XXX_DISPOFF, NULL, 0);
+	if (r < 0) {
+		return r;
+	}
+
+	LOG_HEXDUMP_DBG(regs->pwctrlb, ILI9341_PWCTRLB_LEN, "PWCTRLB");
+	r = ili9xxx_transmit(dev, ILI9341_PWCTRLB, regs->pwctrlb, ILI9341_PWCTRLB_LEN);
+	if (r < 0) {
+		return r;
+	}
```

Comment:
These repeated calls could be rolled into a loop:

```
static const struct {
        uint8_t *reg;
        unsigned int reg_len;
        const char *reg_name;
        uint8_t reg_id;
} transmit_regs[] = {
        { regs->pwctrlb, ILI9341_PWCTRLB_LEN, "PWCTRLB", ILI9341_PWRCTRLB },
        ...
};

for (int i = 0; i < ARRAY_SIZE(transmit_regs); i++) {
        LOG_HEXDUMP_DBG(transmit_regs[i].reg, transmit_regs[i].reg_len, transmit_regs[i].reg_name);
        r = ili9xxx_transmit(dev, transmit_regs[i].reg_id, transmit_regs[i].reg, transmit_regs[i].reg_len);
        if (r < 0) {
                return r;
        }
}
```

Reply:
regs->pwctrlb - not a compile-time constant
fill the array at runtime is a more readable solution, but it also increases memory usage

Reply:
> regs->pwctrlb - not a compile-time constant
> fill the array at runtime is a more readable solution, but it also increases memory usage

The register values are a compile time constant:
    static const struct ili9341_regs ili9xxx_regs_##n

I actually wouldn't expect much difference in the code size for the repeated ili9xxx_transmit calls compared to a loop.

Reply:
I think it allows for more flexibility and easier to read not being in a loop. If there is a register that needs time to settle before setting the next register having the initialization inline allows for a very clear place to put this delay.

Reply:
```
	const struct ili9xxx_config *config = dev->config;
	const struct ili9341_regs *regs = config->regs;
	int r;
	const struct {
		const uint8_t *reg;
		unsigned int reg_len;
		const char *reg_name;
		uint8_t reg_id;
	} transmit_regs[] = {
		{ NULL, 0, "SW RESET", ILI9341_SWRESET },
		{ NULL, 0, "DISPOFF", ILI9XXX_DISPOFF },
		{ regs->pwseqctrl, ILI9341_PWSEQCTRL_LEN, "PWSEQCTRL", ILI9341_PWSEQCTRL },
		{ regs->timctrla, ILI9341_TIMCTRLA_LEN, "TIMCTRLA", ILI9341_TIMCTRLA },
		{ regs->timctrlb, ILI9341_TIMCTRLB_LEN, "TIMCTRLB", ILI9341_TIMCTRLB },
		{ regs->pwctrla, ILI9341_PWCTRLA_LEN, "PWCTRLA", ILI9341_PWCTRLA },
		{ regs->pwctrlb, ILI9341_PWCTRLB_LEN, "PWCTRLB", ILI9341_PWCTRLB },
		{ regs->pumpratio, ILI9341_PUMPRATIOCTRL_LEN, "PUMPRATIOCTRL",
		  ILI9341_PUMPRATIOCTRL },
		{ regs->gamset, ILI9341_GAMSET_LEN, "GAMSET", ILI9341_GAMSET },
		{ regs->frmctr1, ILI9341_FRMCTR1_LEN, "FRMCTR1", ILI9341_FRMCTR1 },
		{ regs->disctrl, ILI9341_DISCTRL_LEN, "DISCTRL", ILI9341_DISCTRL },
		{ regs->pwctrl1, ILI9341_PWCTRL1_LEN, "PWCTRL1", ILI9341_PWCTRL1 },
		{ regs->pwctrl2, ILI9341_PWCTRL2_LEN, "PWCTRL2", ILI9341_PWCTRL2 },
		{ regs->vmctrl1, ILI9341_VMCTRL1_LEN, "VMCTRL1", ILI9341_VMCTRL1 },
		{ regs->vmctrl2, ILI9341_VMCTRL2_LEN, "VMCTRL2", ILI9341_VMCTRL2 },
		{ regs->pgamctrl, ILI9341_PGAMCTRL_LEN, "PGAMCTRL", ILI9341_PGAMCTRL },
		{ regs->ngamctrl, ILI9341_NGAMCTRL_LEN, "NGAMCTRL", ILI9341_NGAMCTRL },
		{ regs->en3g, ILI9341_EN3G_LEN, "EN3G", ILI9341_EN3G },
		{ regs->entrymdset, ILI9341_ETMOD_LEN, "ETMOD", ILI9341_ETMOD },
	};

	for (int i = 0; i < ARRAY_SIZE(transmit_regs); i++) {
		LOG_HEXDUMP_DBG(transmit_regs[i].reg, transmit_regs[i].reg_len,
				transmit_regs[i].reg_name);

		r = ili9xxx_transmit(dev, transmit_regs[i].reg_id, transmit_regs[i].reg,
				     transmit_regs[i].reg_len);
		if (r < 0) {
			return r;
		}
		if (i == 0) {
			k_sleep(K_MSEC(5));
		}
	}
```
transmit_regs - is created on the stack and does not increase the RAM usage in the main program